src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts

```diff
@@ -526,76 +523,90 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 		this._showOutput();
 	}
 
-	private async _executeTask(task: Task, resolver: ITaskResolver, trigger: string, encounteredDependencies: Set<string>, alreadyResolved?: Map<string, string>): Promise<ITaskSummary> {
-		if (encounteredDependencies.has(task.getCommonTaskId())) {
-			this._showDependencyCycleMessage(task);
-			return {};
-		}
-
+	private _executeTask(task: Task, resolver: ITaskResolver, trigger: string, liveDependencies: Set<string>, encounteredTasks: Map<string, Promise<ITaskSummary>>, alreadyResolved?: Map<string, string>): Promise<ITaskSummary> {
 		this._showTaskLoadErrors(task);
 
-		alreadyResolved = alreadyResolved ?? new Map<string, string>();
-		const promises: Promise<ITaskSummary>[] = [];
-		if (task.configurationProperties.dependsOn) {
-			for (const dependency of task.configurationProperties.dependsOn) {
-				const dependencyTask = await resolver.resolve(dependency.uri, dependency.task!);
-				if (dependencyTask) {
-					this._adoptConfigurationForDependencyTask(dependencyTask, task);
-					const key = dependencyTask.getMapKey();
-					let promise = this._activeTasks[key] ? this._getDependencyPromise(this._activeTasks[key]) : undefined;
-					if (!promise) {
-						this._fireTaskEvent(TaskEvent.create(TaskEventKind.DependsOnStarted, task));
-						encounteredDependencies.add(task.getCommonTaskId());
-						promise = this._executeDependencyTask(dependencyTask, resolver, trigger, encounteredDependencies, alreadyResolved);
-					}
-					promises.push(promise);
-					if (task.configurationProperties.dependsOrder === DependsOrder.sequence) {
-						const promiseResult = await promise;
-						if (promiseResult.exitCode === 0) {
-							promise = Promise.resolve(promiseResult);
+		const mapKey = task.getMapKey();
+
+		// It's important that we add this task's entry to _activeTasks before
+		// any of the code in the then runs (see #180541 and #180578). Wrapping
+		// it in Promise.resolve().then() ensures that.
+		const promise = Promise.resolve().then(async () => {
+			alreadyResolved = alreadyResolved ?? new Map<string, string>();
+			const promises: Promise<ITaskSummary>[] = [];
+			if (task.configurationProperties.dependsOn) {
+				liveDependencies = new Set(liveDependencies).add(task.getCommonTaskId());
+				for (const dependency of task.configurationProperties.dependsOn) {
+					const dependencyTask = await resolver.resolve(dependency.uri, dependency.task!);
+					if (dependencyTask) {
+						this._adoptConfigurationForDependencyTask(dependencyTask, task);
+						let promise;
```

Comment:
`promise` is redeclared here (also declared on line 534). It might be clearer to use another variable name.